app/src/main/java/org/simple/clinic/widgets/qrcodescanner/QrCodeAnalyzer.kt

```diff
@@ -0,0 +1,74 @@
+package org.simple.clinic.widgets.qrcodescanner
+
+import android.graphics.ImageFormat.YUV_420_888
+import android.graphics.ImageFormat.YUV_422_888
+import android.graphics.ImageFormat.YUV_444_888
+import android.os.Build
+import androidx.camera.core.ImageAnalysis
+import androidx.camera.core.ImageProxy
+import com.google.zxing.BarcodeFormat
+import com.google.zxing.BinaryBitmap
+import com.google.zxing.DecodeHintType
+import com.google.zxing.MultiFormatReader
+import com.google.zxing.NotFoundException
+import com.google.zxing.PlanarYUVLuminanceSource
+import com.google.zxing.common.HybridBinarizer
+import java.nio.ByteBuffer
+
+private typealias OnQrCodeDetected = (String) -> Unit
+
+class QrCodeAnalyzer(
+    private val onQrCodeDetected: OnQrCodeDetected
+) : ImageAnalysis.Analyzer {
+
+  private fun ByteBuffer.toByteArray(): ByteArray {
+    rewind()
+    val data = ByteArray(remaining())
+    get(data)
+    return data
+  }
+
+  private val reader = MultiFormatReader().apply {
+    val map = mapOf(
+        Pair(DecodeHintType.POSSIBLE_FORMATS, arrayListOf(BarcodeFormat.QR_CODE))
+    )
+    setHints(map)
+  }
+
+  override fun analyze(image: ImageProxy) {
+    val yuvFormats = mutableListOf(YUV_420_888)
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+      yuvFormats.addAll(listOf(YUV_422_888, YUV_444_888))
+    }
```

Comment:
We can extract this into a class property since it will be constant and never change.